src/Data/Array/Accelerate/Trafo/Sharing.hs

```diff
@@ -2044,53 +2044,47 @@ nodeName (AccNodeCount (StableSharingAcc (StableNameHeight sn _) _) _) = NodeNam
 nodeName (ExpNodeCount (StableSharingExp (StableNameHeight sn _) _) _) = NodeName sn
 -- nodeName (SeqNodeCount (StableSharingSeq (StableNameHeight sn _) _) _) = NodeName sn
 
+
 -- Combine node counts that belong to the same node.
 --
 -- * We assume that the list invariant —subterms follow their parents— holds for both arguments and
 --   guarantee that it still holds for the result.
 -- * In the same manner, we assume that all 'Exp' node counts precede 'SmartAcc' node counts and
 --   guarantee that this also hold for the result.
 --
--- RCE: The list combination should be able to be performed as a more efficient merge.
---
+-- With the latest revision, the commented Seq code hasn't been updated for this function.
+
 (+++) :: NodeCounts -> NodeCounts -> NodeCounts
-(ns1,g1) +++ (ns2,g2) = (foldr insert ns1 ns2, Map.unionWith Set.union g1 g2)
+(ns1, g1) +++ (ns2, g2) = (cleanup $ merge ns1 ns2, Map.unionWith Set.union g1 g2)
   where
-    insert x               []                         = [x]
-    insert x@(AccNodeCount sa1 count1) ys@(y@(AccNodeCount sa2 count2) : ys')
-      | sa1 == sa2          = AccNodeCount (sa1 `pickNoneAvar` sa2) (count1 + count2) : ys'
-      | sa1 `higherSSA` sa2 = x : ys
-      | otherwise           = y : insert x ys'
-    insert x@(ExpNodeCount se1 count1) ys@(y@(ExpNodeCount se2 count2) : ys')
-      | se1 == se2          = ExpNodeCount (se1 `pickNoneVar` se2) (count1 + count2) : ys'
-      | se1 `higherSSE` se2 = x : ys
-      | otherwise           = y : insert x ys'
-    -- insert x@(SeqNodeCount se1 count1) ys@(y@(SeqNodeCount se2 count2) : ys')
-    --   | se1 == se2          = SeqNodeCount (se1 `pickNoneSvar` se2) (count1 + count2) : ys'
-    --   | se1 `higherSSS` se2 = x : ys
-    --   | otherwise           = y : insert x ys'
-    insert x@(AccNodeCount _ _) (y@(ExpNodeCount _ _) : ys')
-      = y : insert x ys'
-    insert x@(ExpNodeCount _ _) (y@(AccNodeCount _ _) : ys')
-      = x : insert y ys'
```

Comment:
This is the line that caused cubic behaviour: replacing it with the below passed doctests:
`x : y : ys'`